### PR TITLE
[Team Offline Merge] updates models + created and updates routes

### PR DESCRIPTION
**Pull request by sofiiagran**
## Description
last updated 19:57 12/10/23 : added extra error handling as well as testing error handling

We have changed and created some new routes and updated models

## Changes

We have updated models so they follow the structure on Miro.
We have commented out all the routes we are not using.

The new routes are found in courseRoutes and are related to fetching courses and sections, as well as subscribing to course.

We also did some modifications in the authRoute so that the user id is stored in addition to mail and name from before.

## Related Issues

Not really.

## Checklist

- [ yes ] Code has been tested locally and passes all relevant tests.
- [ yes ] Documentation has been updated to reflect the changes, if applicable.
- [ yes ] Code follows the established coding style and guidelines of the project.
- [ yes ] All new and existing tests related to the changes have passed.
- [ yes ] Any necessary dependencies or new packages have been properly documented.
- [ yes ] Pull request title and description are clear and descriptive.
- [ yes ] Reviewers have been assigned to the pull request.
- [ yes ] Any potential security implications have been considered and addressed.
- [ yes ] Performance impact of the changes has been evaluated, if relevant.

## Notes for Reviewers

Changed the naming of some .spec files in the application / content-creator application folder, since the tests are from last year and all the tests fail. they are now named with notInUseTest_ at the beginning and removed the .specs to avoid testing them for now.

All routes have been tested in the __test__ folder

### CODE REVIEW

File: __tests__/routes/courseRoutes.specs.js
Comment (Mafusn) [inline __tests__/routes/courseRoutes.specs.js:1]: Burde hedde courseRoutes.spec.js, ellers fint

File: applications/content-creator-applications/controller/contentCreatorApplicationControllerspec.js
Comment (Mafusn) [inline applications/content-creator-applications/controller/contentCreatorApplicationControllerspec.js:1]: Why are this renamed?

File: applications/content-creator-applications/data-access/contentCreatorApplicationListspec.js
Comment (Mafusn) [inline applications/content-creator-applications/data-access/contentCreatorApplicationListspec.js:1]: Why is this renamed, i don't think it runs when we call the script test if it is not named .spec.js

File: routes/courseRoutes.js
Comment (Mafusn) [inline routes/courseRoutes.js:1]: I know it's not you, but maybe add "/courses" to index.js's course route in the route folder to prevent the repetition of "courses" in the courseRoutes.js file

**Review by Mafusn — CHANGES_REQUESTED** [review]
Looks good, only minor changes

File: __tests__/routes/courseRoutes.specs.js
Comment (sofiiagran) [reply __tests__/routes/courseRoutes.specs.js:1]: I'm I blind but is it not named courseRoutes.spec.js? and the indention error is only in github, not when i open the file, idk why?

File: applications/content-creator-applications/controller/contentCreatorApplicationControllerspec.js
Comment (sofiiagran) [reply applications/content-creator-applications/controller/contentCreatorApplicationControllerspec.js:1]: its from last year and all the tests fail

Comment (sofiiagran) [reply applications/content-creator-applications/controller/contentCreatorApplicationControllerspec.js:1]: its just annoying to run tests on it

File: routes/courseRoutes.js
Comment (sofiiagran) [reply routes/courseRoutes.js:1]: I have fixed it and moved the /users routes to the userRoutes.js

File: applications/content-creator-applications/data-access/contentCreatorApplicationListspec.js
Comment (sofiiagran) [reply applications/content-creator-applications/data-access/contentCreatorApplicationListspec.js:1]: This is tests from last year and all of them are failing, I know renamed it again to notInUseTest_contectCreatorApplication.js so that it is not at spec and the tests are not ran

File: __tests__/routes/courseRoutes.specs.js
Comment (sofiiagran) [reply __tests__/routes/courseRoutes.specs.js:1]: fixed

File: routes/userRoutes.js [hunk 1]
```diff
@@ -88,6 +89,56 @@ async function validateFields(fields) {
...
+// Get users subscriptions
+router.get('/:id/subscriptions', async (req, res) => {
+  try {
+    const userId = req.params.id;
+
+    // Find the user by _id and select the 'subscriptions' field
+    const user = await User.findById(userId).select('subscriptions');
+
+    if (!user) {
+      return res.status(404).json({ message: 'User not found' });
```
Comment (Mafusn) [inline routes/userRoutes.js:103]: Should return error 204 as stated in microsoft API guidelines (error 204 can't return a response)

File: __tests__/routes/courseRoutes.spec.js
Comment (Mafusn) [inline __tests__/routes/courseRoutes.spec.js:1]: Needs some tests for handling errors

**Review by Mafusn — CHANGES_REQUESTED** [review]
Looks good, only needs some more test cases for errors, and maybe take a look at the file errorCodes.js in the helpers folder. It would be great to implement them as soon as possible, there is a readme file in the resources repo for them. Quick note, why are you adding an .idea folder.

File: __tests__/routes/courseRoutes.spec.js
Comment (sofiiagran) [reply __tests__/routes/courseRoutes.spec.js:1]: :((((

File: routes/userRoutes.js (hunk 1 above)
Comment (sofiiagran) [reply routes/userRoutes.js]: image

Comment (sofiiagran) [reply routes/userRoutes.js]: ...?

Comment (sofiiagran) [reply routes/userRoutes.js]: Jeg tror det kun gælder for DELETE requests. Jeg spurgte chatGPT for at dobbelttjekke:

Yes, using a 404 status code in this context is appropriate, assuming you are following RESTful principles for your microservices API. In your code, you're attempting to retrieve a user's subscriptions by their ID. If the user with the given ID doesn't exist, returning a 404 status code is a valid response.

HTTP status code 404 (Not Found) is used to indicate that the requested resource (in this case, the user) was not found on the server. It's a standard way to communicate to the client that the requested data doesn't exist.

So, your code is correctly using a 404 status code to indicate that the user was not found, and it provides a relevant error message in the JSON response ("User not found"). This is in line with RESTful API design best practices.

File: __tests__/routes/courseRoutes.spec.js
Comment (sofiiagran) [reply __tests__/routes/courseRoutes.spec.js:1]: har lavet sindssygt mange tests for handling errors

**Comment by sofiiagran** [comment]
lots of tests and error handling added :))

File: routes/courseRoutes.js [hunk 2]
```diff
@@ -437,4 +452,245 @@ router.get('/course/delete_all', requireLogin, async (req, res) => {
...
+	res.status(500).json({ message: 'Server error' });
+	}
+
+});
+*/
+
+/*** SUBSCRIPTION ROUTES ***/
+
+// Subscribe to course 
+router.post('/:id/subscribe', async (req, res) => {
```
Comment (Havkost) [inline routes/courseRoutes.js:623]: Right now our [api guidelines](https://github.com/Educado-App/resources/blob/main/API_GUIDELINES.md#performing-an-action) specify that actions like subscribe should be formatted as e.g. `/couses/some_id:subscribe`.

We should either change the guidelines or follow them. Maybe this is a question for the integration team?

How to add colon in express endpoints:
https://github.com/expressjs/express/issues/3857#issuecomment-589792737

File: routes/courseRoutes.js
Comment (Havkost) [inline routes/courseRoutes.js:1]: Much of this file is commented out, can we just remove those parts?

**Review by Havkost — APPROVED** [review]
## Approved
Great job, especially on the tests 👍

We should discuss how we format actions in our API routes, but I don't think that discussion should block a merge now.

File: routes/courseRoutes.js (hunk 2 above)
Comment (sofiiagran) [reply routes/courseRoutes.js:623]: it is /courses, check index.js

File: routes/courseRoutes.js
Comment (sofiiagran) [reply routes/courseRoutes.js:1]: yes i would say so, we are not using it, it is from last year

File: routes/courseRoutes.js (hunk 2 above)
Comment (sofiiagran) [reply routes/courseRoutes.js:623]: The second option courses/id:subscribe doesn't follow the standard RESTful URL structure, and it might be confusing for other developers who are used to working with RESTful APIs. It also doesn't provide an obvious way to capture the id of the course.

So, in most cases, it's better to stick with the first option (courses/:id/subscribe) for defining routes in a microservices API following RESTful principles.

Comment (sofiiagran) [reply routes/courseRoutes.js:623]: jeg spurgte lige chat GPT: 

" hello, following microservices API guidelines, should I write my route like: router.post('courses/:id/subscribe', async (req, res) => {                                                                                                          
     // or should i do it like this:                                                                                                                                                            router.post('courses/id:subscribe', async (req, res) => {    "

In the context of designing a RESTful API for microservices, the first option you provided is the recommended way to define routes. Here's why:

RESTful URL Structure: REST (Representational State Transfer) is an architectural style that uses a specific URL structure to represent resources and actions. In a RESTful API, resources are typically represented by nouns (like "courses" in your example), and actions are represented by HTTP methods (like POST for creating a subscription). Using the URL path segment courses/:id/subscribe aligns with REST principles, where :id is a placeholder for the specific course's identifier. This makes your API more self-explanatory and easier to understand for developers.

Parameter Parsing: In most web frameworks or libraries (e.g., Express.js for Node.js), you can easily extract the :id parameter from the URL. It allows you to access the id in your route handler, which is crucial for performing actions specific to a course.

Consistency: Following standard conventions, such as RESTful URLs, makes your API consistent and predictable. Developers who work with your API will already be familiar with these conventions and can understand how to interact with it more easily.

The second option courses/id:subscribe doesn't follow the standard RESTful URL structure, and it might be confusing for other developers who are used to working with RESTful APIs. It also doesn't provide an obvious way to capture the id of the course.

So, in most cases, it's better to stick with the first option (courses/:id/subscribe) for defining routes in a microservices API following RESTful principles.

Comment (sofiiagran) [reply routes/courseRoutes.js:623]: Jeg tror ikke subscribe er "action" som beskrivet i API guidelines, siden den opdaterer modellen til user: defor tror jeg det går under denne definition: 
[⛔](https://github.com/Educado-App/resources/blob/main/API_GUIDELINES.md#actions-no-actions-for-crud) DO NOT use an action operation when the operation behavior could reasonably be defined as one of the standard REST Create, Read, Update, Delete, or List operations.

Comment (sofiiagran) [reply routes/courseRoutes.js:623]: altså subscribe tror jeg ikke er action som beskrevet her: [Performing an Action](https://github.com/Educado-App/resources/blob/main/API_GUIDELINES.md#performing-an-action)
The REST specification is used to model the state of a resource, and is primarily intended to handle CRUD (Create, Read, Update, Delete) operations. However, many services require the ability to perform an action on a resource, e.g. getting the thumbnail of an image or rebooting a VM. It is also sometimes useful to perform an action on a collection.

Comment (sofiiagran) [reply routes/courseRoutes.js:623]: og chatGPT mente at courses/:id/subscribe var rigtigt

File: helpers/errorCodes.js [hunk 3]
```diff
@@ -21,6 +21,26 @@ module.exports = {
...
+		code: 'E0005',
+		message: 'No courses found'
+	},
+	E0006: {
+		code: 'E0006',
+		message: 'Course not found'
+	},
+	E0007: {
+		code: 'E0007',
+		message: 'No sections not found'
```
Comment (Mafusn) [inline helpers/errorCodes.js:34]: "No sections not found" maybe "No sections found"

File: routes/userRoutes.js [hunk 4]
```diff
@@ -88,6 +89,75 @@ async function validateFields(fields) {
...
+    // If the server could not be reached, return an error message
+    console.log(error);
+    return res.status(500).json({ 'error': errorCodes['E0003'] });
+  }
+});
+
+
+
+// Checks if user is subscribed to a specific course
+router.get('', async (req, res) => {
```
Comment (Mafusn) [inline routes/userRoutes.js:125]: Is this route correct?

**Review by Mafusn — CHANGES_REQUESTED** [review]
Only two very small comments

File: helpers/errorCodes.js (hunk 3 above)
Comment (sofiiagran) [reply helpers/errorCodes.js]: fixed

File: routes/userRoutes.js (hunk 4 above)
Comment (sofiiagran) [reply routes/userRoutes.js]: fixed

**Review by Mafusn — APPROVED** [review]
Great work!
